crates/bevy_ecs/src/world/mod.rs

```diff
@@ -0,0 +1,819 @@
+    component::{
+        Component, ComponentDescriptor, ComponentFlags, ComponentId, Components, ComponentsError,
+        StorageType,
+    },
+    entity::{Entities, Entity},
+    query::{FilterFetch, QueryState, WorldQuery},
+    storage::{Column, SparseSet, Storages},
+};
+use std::{any::TypeId, fmt};
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct WorldId(u64);
+
+impl Default for WorldId {
+    fn default() -> Self {
+        WorldId(rand::random())
+    }
+}
+
+/// [World] stores and exposes operations on [entities](Entity), [components](Component), and their associated metadata.
+/// Each [Entity] has a set of components. Each component can have up to one instance of each component type.
+/// Entity components can be created, updated, removed, and queried using a given [World].
+#[derive(Default)]
+pub struct World {
+    id: WorldId,
+    pub(crate) entities: Entities,
+    pub(crate) components: Components,
+    pub(crate) archetypes: Archetypes,
+    pub(crate) storages: Storages,
+    pub(crate) bundles: Bundles,
+    pub(crate) removed_components: SparseSet<ComponentId, Vec<Entity>>,
+    pub(crate) archetype_component_access: ArchetypeComponentAccess,
+    main_thread_validator: MainThreadValidator,
+}
+
+impl World {
+    /// Creates a new empty [World]
+    #[inline]
+    pub fn new() -> World {
+        World::default()
+    }
+
+    /// Retrieves this world's unique ID
+    #[inline]
+    pub fn id(&self) -> WorldId {
+        self.id
+    }
+
+    /// Retrieves this world's [Entities] collection
+    #[inline]
+    pub fn entities(&self) -> &Entities {
+        &self.entities
+    }
+
+    /// Retrieves this world's [Archetypes] collection
+    #[inline]
+    pub fn archetypes(&self) -> &Archetypes {
+        &self.archetypes
+    }
+
+    /// Retrieves this world's [Components] collection
+    #[inline]
+    pub fn components(&self) -> &Components {
+        &self.components
+    }
+
+    /// Retrieves a mutable reference to this world's [Components] collection
+    #[inline]
+    pub fn components_mut(&mut self) -> &mut Components {
+        &mut self.components
+    }
+
+    /// Retrieves this world's [Storages] collection
+    #[inline]
+    pub fn storages(&self) -> &Storages {
+        &self.storages
+    }
+
+    /// Retrieves this world's [Bundles] collection
+    #[inline]
+    pub fn bundles(&self) -> &Bundles {
+        &self.bundles
+    }
+
+    /// Retrieves a [WorldCell], which safely enables multiple mutable World accesses at the same time,
+    /// provided those accesses do not conflict with each other.
+    #[inline]
+    pub fn cell(&mut self) -> WorldCell<'_> {
+        WorldCell::new(self)
+    }
+
+    /// Registers a new component using the given [ComponentDescriptor]. Components do not need to be manually
+    /// registered. This just provides a way to override default configuration. For example, the default
+    /// component storage type can be overridden like this:
```

Comment:
Maybe add something about that the component type must not have been previously used.